lib/libshare/os/linux/nfs.c

```diff
@@ -226,66 +233,57 @@ get_linux_hostspec(const char *solaris_hostspec, char **plinux_hostspec)
 }
 
 /*
- * Used internally by nfs_enable_share to enable sharing for a single host.
+ * This function populates an entry into /etc/exports.d/zfs.exports.
+ * This file is consumed by the linux nfs server so that zfs shares are
+ * automatically exported upon boot or whenever the nfs server restarts.
  */
 static int
-nfs_enable_share_one(const char *sharepath, const char *host,
-    const char *security, const char *access, void *pcookie)
+nfs_add_entry(const char *filename, const char *sharepath,
+    const char *host, const char *security, const char *access_opts,
+    void *pcookie)
 {
 	int rc;
-	char *linuxhost, *hostpath, *opts;
+	char *linuxhost;
 	const char *linux_opts = (const char *)pcookie;
-	char *argv[6];
-
-	/* exportfs -i -o sec=XX,rX,<opts> <host>:<sharepath> */
 
 	rc = get_linux_hostspec(host, &linuxhost);
+	if (rc != SA_OK)
+		return (rc);
 
-	if (rc < 0)
-		exit(1);
-
-	hostpath = malloc(strlen(linuxhost) + 1 + strlen(sharepath) + 1);
+	if (linux_opts == NULL)
+		linux_opts = "";
 
-	if (hostpath == NULL) {
+	const char *fmt = "%s %s(sec=%s,%s,%s)\n";
+	int bytes = snprintf(NULL, 0, fmt, sharepath,
+	    linuxhost, security, access_opts, linux_opts);
+	char *buffer = malloc(bytes + 1);
+	if (buffer == NULL) {
 		free(linuxhost);
-
-		exit(1);
+		return (SA_NO_MEMORY);
 	}
 
-	sprintf(hostpath, "%s:%s", linuxhost, sharepath);
-
+	sprintf(buffer, fmt, sharepath, linuxhost,
+	    security, access_opts, linux_opts);
 	free(linuxhost);
 
-	if (linux_opts == NULL)
-		linux_opts = "";
-
-	opts = malloc(4 + strlen(security) + 4 + strlen(linux_opts) + 1);
-
-	if (opts == NULL)
-		exit(1);
-
-	sprintf(opts, "sec=%s,%s,%s", security, access, linux_opts);
-
-#ifdef DEBUG
-	fprintf(stderr, "sharing %s with opts %s\n", hostpath, opts);
-#endif
-
-	argv[0] = "/usr/sbin/exportfs";
-	argv[1] = "-i";
-	argv[2] = "-o";
-	argv[3] = opts;
-	argv[4] = hostpath;
-	argv[5] = NULL;
-
-	rc = libzfs_run_process(argv[0], argv, 0);
-
-	free(hostpath);
-	free(opts);
+	int fd = open(filename, O_RDWR | O_APPEND | O_CREAT, 0644);
+	if (fd == -1) {
+		fprintf(stderr, "failed to open %s file: %s", filename,
+		    strerror(errno));
+		free(buffer);
+		return (SA_SYSTEM_ERR);
+	}
 
-	if (rc < 0)
+	if (write(fd, buffer, strlen(buffer)) == -1) {
+		perror("failed to write exports file");
+		close(fd);
+		free(buffer);
 		return (SA_SYSTEM_ERR);
-	else
-		return (SA_OK);
+	}
+	fsync(fd);
```

Comment:
Why do we need fo fsync() here, rather than in nfs_export_all()?  Seems like it would perform better to do it just once per logical operation.  If we crash in the middle, before nfs_export_all(), it seems like the file would be half-modified even with this fsync().